code/parse/sexp.cpp

```diff
@@ -10561,6 +10566,62 @@ int eval_for_ship_collection(int arg_handler_node, int condition_node, int op_co
 		return SEXP_FALSE;
 }
 
+// jg18
+int eval_for_container(int arg_handler_node, int condition_node, int op_const, bool just_count = false)
+{
+	Assertion(arg_handler_node != -1,
+		"Attempt to use invalid argument handler with a for-container SEXP (%d). Please report!",
+		op_const);
+	Assertion(condition_node != -1,
+		"Attempt to use invalid condition with a for-container SEXP (%d). Please report!",
+		op_const);
+
+	int num_arguments = 0;
+	SCP_vector<std::pair<char*, int>> argument_vector;
+
+	const int arg_node = CDR(arg_handler_node);
+
+	switch (op_const)
+	{
+		case OP_FOR_CONTAINER_DATA:
+			num_arguments = sexp_container_collect_data_arguments(arg_node, argument_vector, just_count);
+			break;
+
+		case OP_FOR_MAP_CONTAINER_KEYS:
+			num_arguments = sexp_container_collect_map_key_arguments(arg_node, argument_vector, just_count);
+			break;
+
+		default:
+			UNREACHABLE("Unhandled for-container SEXP (%d). Please report!", op_const);
+			break;
+	}
+
+	if (just_count) {
+		// cleaning up argument_vector isn't needed
+		// because the "collect arguments" functions didn't touch it
+		return num_arguments;
+	}
+
+	// test the whole argument vector
+	int num_true, num_false, num_known_true, num_known_false;
+	const int num_valid_arguments = test_argument_vector_for_condition(argument_vector,
+		true,
+		condition_node,
+		&num_true,
+		&num_false,
+		&num_known_true,
+		&num_known_false);
+
+	// use the sexp_or algorithm
+	if (num_known_true || num_true)
+		return SEXP_TRUE;
+	// TODO: is the check below correct?
+	else if (num_valid_arguments > 0 && num_known_false == num_valid_arguments)
+		return SEXP_KNOWN_FALSE;
```

Comment:
Is this `else-if` clause correct, in terms of fulfilling expected behavior? Or should it be commented out, like in `eval_for_ship_collection()`?

Reply:
I am not sure, but I don't think it is correct for the container stuff to return ``SEXP_KNOWN_FALSE``, because (if I understood correctly) it is possible that an event might _add_ a value to a container. Which in turn means that an event that evaluated ``KNOWN_FALSE`` for all valid arguments (and thus would flag here as ``KNOWN_FALSE`` as well) is stopped being evaluated, even with the new value added. To prevent that we must only return ``SEXP_FALSE``, so that it is reevaluated when new items are added.